### PR TITLE
fix #19500; remove find optimization [backport: 1.6]

### LINKED ISSUE #19500: `strutils.find` uses cstring optimization that stops after \0
### Example
```nim
import strutils

echo "abc \0 def".find("def")
```

### Current Output
```
-1
```

### Expected Output
```
6
```

### Possible Solution

* Remove [this cstring optimization in strutils.find](https://github.com/nim-lang/Nim/blob/6b302bb71b6d9aab708b91f8d68957652ec30480/lib/pure/strutils.nim#L1875-L1880) and maybe move it as an overload to `cstrutils`.

### Additional Information

* From forum thread https://forum.nim-lang.org/t/8874

```
$ nim -v
Nim Compiler Version 1.6.2
```

### PR DESCRIPTION
**Pull request by ringabout**
fix https://github.com/nim-lang/Nim/issues/19500
close https://github.com/nim-lang/Nim/pull/19504

### CODE REVIEW

File: lib/std/private/strimpl.nim [hunk 1]
```diff
@@ -74,3 +74,40 @@ template endsWithImpl*[T: string | cstring](s, suffix: T) =
...
   cmpIgnoreStyleImpl(a, b, true)
+
+func c_memchr(cstr: pointer, c: char, n: csize_t): pointer {.
+              importc: "memchr", header: "<string.h>".}
+func c_strstr(haystack, needle: cstring): cstring {.
+  importc: "strstr", header: "<string.h>".}
+
+
+func find*(s: cstring, sub: char, start: Natural = 0, last = 0): int =
+  ## Searches for `sub` in `s` inside range `start..last` (both ends included).
```
Comment (konsumlamm) [inline lib/std/private/strimpl.nim]:
```suggestion
  ## Searches for `sub` in `s` inside the range `start..last` (both ends included).
```

File: lib/std/private/strimpl.nim [hunk 2]
```diff
@@ -74,3 +74,40 @@ template endsWithImpl*[T: string | cstring](s, suffix: T) =
...
+  let last = if last == 0: s.high else: last
+  let L = last-start+1
+  if L > 0:
+    let found = c_memchr(s[start].unsafeAddr, sub, cast[csize_t](L))
+    if not found.isNil:
+      return cast[ByteAddress](found) -% cast[ByteAddress](s)
+  return -1
+
+func find*(s, sub: cstring, start: Natural = 0, last = 0): int =
+  ## Searches for `sub` in `s` inside range `start..last` (both ends included).
```
Comment (konsumlamm) [inline lib/std/private/strimpl.nim]:
```suggestion
  ## Searches for `sub` in `s` inside the range `start..last` (both ends included).
```

**Review by konsumlamm — CHANGES_REQUESTED** [review]

**Comment by Varriount** [comment]
Just to ensure I understand the changes correctly, how is this different from https://github.com/nim-lang/Nim/pull/19504 ?

File: lib/std/private/strimpl.nim [hunk 3]
```diff
@@ -74,3 +74,40 @@ template endsWithImpl*[T: string | cstring](s, suffix: T) =
...
+  ## Searching is case-sensitive. If `sub` is not in `s`, -1 is returned.
+  ## Otherwise the index returned is relative to `s[0]`, not `start`.
+  ## Use `s[start..last].rfind` for a `start`-origin index.
+  let last = if last == 0: s.high else: last
+  let L = last-start+1
+  if L > 0:
+    let found = c_memchr(s[start].unsafeAddr, sub, cast[csize_t](L))
+    if not found.isNil:
+      return cast[ByteAddress](found) -% cast[ByteAddress](s)
+  return -1
```
Comment (ringabout) [inline lib/std/private/strimpl.nim:97]: Not much, I just kept changes in the private modules.

**Review by Varriount — APPROVED** [review]

**Comment by ringabout** [comment]
Since it has been merged, I will use memmem to optimize it later.

**Pull request merged**